### ISSUE TITLE
Delayed Parameter Update when step(wait=False)

### ISSUE THREAD

**Issue by bgyoon** [learning-at-home/hivemind#394]
**Is your feature request related to a problem? Please describe.**

Eh, this could be a question. I'm trying to use `TrainingAverager` with `step(wait=False)`. That requires `data_lock` and `use_old_local_tensor=True` follows.

When `use_old_local_tensor=True`, is it correct to simply add the weight difference between local model and all-reduced model to the new model parameters? The gradients calculated from the old model parameter is being added to the new model parameters. That doesn't seem quite right.

**Describe the solution you'd like**

https://arxiv.org/abs/2101.06840 proposes Delayed Parameter Update. Parameter update is delayed by one step. Apparently, it makes little difference in the training curve if DPU is applied after 40 iterations in BERT-large training.

I think to implement DPU, you simply have to copy back the averaged tensor back to the model in the beginning of `step()`.

**Describe alternatives you've considered**

I understand that if the weight difference is not added back, the local steps taken before  the asynchronous all-reduce completes are being wasted. Not only it defeats purpose of asynchronous all-reduce(if local updates are going to be wasted until async completes, why not just go sync) but it also skips over input data which could trouble training.

**Comment by justheuristic** [comment]
Hi!

What you're saying is perfectly correct, and with proper application, it will be equivalent to [2101.06840](https://arxiv.org/abs/2101.06840).

As you've said, ZeRO-offload is equivalent to at most 1 step of staleness. Therefore, there won't be any asynchronous **parameter** updates in the background for all-reduce. Instead, here's what should happen:

- first, peers accumulate gradients until they reach the target batch size
- then, IN PARALLEL
    - THREAD A - average currently accumulated gradients and run optimizer step
    - THREAD B - begin accumulating gradients for the next step - but do not run the next optimizer step yet

Therefore, we implemented DPU without using wait=False, but with two threads:
- thread A runs entirely on CPU, runs TrainingAverager and performs the optimizer steps  
- thread B uses GPU, accumulates gradients into a separate on-gpu buffer

These threads have only two points of interaction: A receives gradients from B before each all-reduce and sends back updated parameters after optimizer step. 

<details>
 <summary>More specifically, this (overly complicated) algorithm</summary>
- [ thread A, without GPU, with TrainingAverager ] 
  1. record local parameters & gradients from GPU as old_local_tensors
  2. run all_reduce over the network, get averaged gradients
      - during this step, the parameters/gradients in thread A will NOT change, thread B does not run optimizer steps
  4. run optimizer on averaged gradients, get updated parameters
  5. push updated parameters to thread B
  6. wait for GPU to accumulate enough gradients for the next step
 - [ thread B, with GPU, with separate gradient buffers ]
   - accumulate gradients perpetually. After each forward/backward pass,
      - if thread A just updated its parameters, also update GPU parameters
      - if accumulated enough gradients, push gradients to thread A 
</details>

**Comment by justheuristic** [comment]
> when lr is decreased by lr scheulder(0.1 times in step-wise fashion) at 1.2kish steps, the training seems to be working

That might indeed be the case. In our DPU experiments, it was enabled during the initial LR warmup, where the learning rate was still very small. That might have allowed DPU to phase in without significant performance drawdown.

p.s. i've finished the rest of my backlog yesterday, now working on making the DPU work in hivemind master. I'd still appreciate if you have time to chat a little to better coordinate our effort. (we can meet on discord or whichever other means of communication you prefer)

**Comment by justheuristic** [comment]
To whom it may concern:
delayed parameter updates are enabled with hivemind.Optimizer via ```delay_grad_averaging=True, delay_optimizer_step=True``` .

Minimalistic example: [benchmark_optimizer.py](https://github.com/learning-at-home/hivemind/blob/master/benchmarks/benchmark_optimizer.py#L30-L31)

More advanced usage examples (with full or partial DPU up to user's discretion):
- https://github.com/learning-at-home/dalle-hivemind
- https://github.com/SeanNaren/hivemind-lightning/blob/main/wikitext.py#L485

A more detailed API reference can be found here:
- https://learning-at-home.readthedocs.io/en/latest/modules/optim.html

Currently, DPU requires installing hivemind from the github repo, i.e.
`pip install https://github.com/learning-at-home/hivemind/archive/master.zip`

It will be available from PyPI after v1.0.0 is [released](https://github.com/learning-at-home/hivemind/releases), which is to say "sometime very soon"

If you have any other questions, feel free to open another issue or join our discord channel (link above)
